core/types/merkle.go

```diff
@@ -0,0 +1,152 @@
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/bits"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+const (
+	Txs int = iota
+	Recs
+	Withd
+)
+
+var lookup map[int]common.Hash
+
+func init() {
+	// initialize the lookup
+	max := 1 << 30
+	base := [32]byte{}
+	prev := crypto.Keccak256Hash(base[:])
+	lookup[1] = prev
+	for i := 2; i < max; i = i << 1 {
+		prev = crypto.Keccak256Hash(prev.Bytes(), prev.Bytes())
+		lookup[i] = prev
+	}
+}
+
+// Merkleize computes a binary merkle tree of a derivable list.
+// It's similar to deriveSha.
+func Merkleize(list DerivableList, typ int) common.Hash {
+	valueBuf := encodeBufferPool.Get().(*bytes.Buffer)
+	defer encodeBufferPool.Put(valueBuf)
+
+	sha := hasherPool.Get().(crypto.KeccakState)
+	defer hasherPool.Put(sha)
+
+	var (
+		innerLimit int
+		outerLimit int
+	)
+	switch typ {
+	case Txs:
+		innerLimit = 1 << 30
+		outerLimit = 1 << 20
+	case Recs:
+		// TODO research limits
+	case Withd:
+		innerLimit = 1 << 2
+		outerLimit = 1 << 4
+	}
+
+	size := nextPowerOfTwo(uint32(list.Len()))
+	roots := make([]common.Hash, size)
+	for i := 0; i < list.Len(); i++ {
+		valueBuf.Reset()
+		list.EncodeIndex(i, valueBuf)
+		bytes := valueBuf.Bytes()
+		roots[i] = mixinLength(chunkAndHashSSZ(bytes, innerLimit), len(bytes))
+	}
+	return mixinLength(merkleWithLimit(roots, outerLimit), list.Len())
+}
+
+// chunkAndHashSSZ cuts an input into chunks of 32 bytes
+// and creates a merkle tree out of them
+func chunkAndHashSSZ(input []byte, limit int) common.Hash {
+	size := nextPowerOfTwo(uint32(len(input)))
+	roots := make([]common.Hash, size)
+	for start := 0; start < len(input); start += common.HashLength {
+		end := start + common.HashLength
+		if end > len(input) {
+			end = len(input)
+		}
+		copy(roots[start/common.HashLength][:], input[start:end])
+	}
+	return merkleWithLimit(roots, limit)
+}
+
+// merkleWithLimit recursively computes a merkle tree with the help of a LUT.
+// The limits are set very liberally which means a lot of the tree can be precomputed.
+// All data is left aligned in the leaves, so we can recursively compute the left children.
+// The right children can be looked up from the table.
+func merkleWithLimit(values []common.Hash, limit int) (h common.Hash) {
```

Comment:
The way you're hashing this you will complete with zero chunks at the base to the next power of two and continuously hash 0 + 0 several times. Say you have 129 chunks to hash, you will complete this to 256 by adding 127 chunks of zero that will be hashed as if they contained data. The right way to hash this is by just hash against the appropriate value of `lookup`. 

If you use gohashtree however it's even better cause you only need to hash once per depth.